server/tilesource/base.py

```diff
@@ -54,6 +59,104 @@ def __init__(self):
         self.sizeX = None
         self.sizeY = None
 
+    def _calculateWidthHeight(self, width, height, sourceWidth, sourceHeight):
+        """
+        Given a source width and height and a maximum destination width and/or
+        height, calculate a destination width and height that preserves the
+        aspect ratio of the source.
+
+        :param width: the destination width.  None to only use height.
+        :param height: the destination height.  None to only use width.
+        :param sourceWidth: the width of the source data.
+        :param sourceHeight: the height of the source data.
+        :returns: the width and height that is no larger than that specified
+                  and preserves aspect ratio.
+        """
+        if sourceWidth == 0 or sourceHeight == 0:
+            return 0, 0
+        # Constrain the maximum size if both width and height weren't
+        # specified, in case the image is very short or very narrow.
+        if height and not width:
+            width = height * 16
```

Comment:
This is only for when the user hasn't specified both dimensions for limitation.  I don't like the idea of unbounded results when the user may not realize that is what they are asking for.  For instance, supposed you have an original image that is extremely short and wide. say 1x16000, and the user asks for a height of 1000 pixels, it seems imprudent to try to generate that.  If they really want it, they can ask for (a very large number)x1000.